File: benchmarks/matmul/matmul.cpp
#include <iostream>
#include <vector>

void matmul(int n, const std::vector<float>& a, const std::vector<float>& b, std::vector<float>& c) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            float sum = 0.0f;
            for (int k = 0; k < n; k++) {
                sum += a[i * n + k] * b[k * n + j];
            }
            c[i * n + j] = sum;
        }
    }
}

int main() {
    int n = 512;
    std::vector<float> a(n * n, 1.0f);
    std::vector<float> b(n * n, 2.0f);
    std::vector<float> c(n * n, 0.0f);
    
    matmul(n, a, b, c);
    
    std::cout << "Matmul result[0] = " << c[0] << std::endl;
    return 0;
}

File: benchmarks/matmul/matmul.rs
fn matmul(n: usize, a: &[f32], b: &[f32], c: &mut [f32]) {
    for i in 0..n {
        for j in 0..n {
            let mut sum = 0.0;
            for k in 0..n {
                sum += a[i * n + k] * b[k * n + j];
            }
            c[i * n + j] = sum;
        }
    }
}

fn main() {
    let n = 512;
    let a = vec![1.0; n * n];
    let b = vec![2.0; n * n];
    let mut c = vec![0.0; n * n];
    
    matmul(n, &a, &b, &mut c);
    
    println!("Matmul result[0] = {}", c[0]);
}
